### PR TITLE
Split core traits

### LINKED ISSUE #460: Split core traits
This has been discussed before, IIRC with @rfuest raising the suggestion that we split core items like `DrawTarget` out into an `embedded-graphics-core` trait. 

I'm opening this issue because it might fix another issue: `tinytga` is used in the e-g doc examples, but `tinytga` depends on e-g itself, creating a cyclic dependency that breaks as soon as the versions aren't _perfectly_ in sync. This is obviously not good, so along with `DrawTarget`, we may want to consider splitting more traits for images and primitives, as well as the pixel colours, into a new core crate. 

If we go ahead with something like this, should it target 0.7 or go for a 0.8 release?

Discuss.

### PR DESCRIPTION
**Pull request by jamwaffles**
Thank you for helping out with embedded-graphics development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) if your changes affect the **public API**
- [x] Run `rustfmt` on the project
- [x] Run `just build` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

This is going to be a horrific PR to review, so sorry for that in advance.

Do we want to re-export `embedded-graphics-core` from e-g? I have a `pub use embedded_graphics_core as core;` at the moment which allows things like `use embedded-graphics::core::pixelcolor::Rgb565;` which isn't awful. Another thing I considered is `pub use embedded_graphics_core::{pixelcolor, geometry, drawable::*};` which would result in fewer changes in user code.

I chose to go with the `csv`-ish structure, with the core as a subfolder of the top level. I think this makes sense as there are fewer changes required.

Am I going in the right direction with this @rfuest? 

Closes #460 

TODO

* [x] Flesh out migration guide just a little bit, to reduce required effort later.
* [x] Add documentation to e-g lib.rs to point driver authors toward using e-g-core
* [x] Find other documentation about implementing e-g-core stuff and refactor
* [x] Add e-g-core lib.rs docs.

### CODE REVIEW

**Comment by rfuest** [comment]
> Another thing I considered is pub use embedded_graphics_core::{pixelcolor, geometry, drawable::*}; which would result in fewer changes in user code.

Yes, this was my plan. IMO the separation into two crates shouldn't normally be visible for users of e-g and only matter for driver, text renderer or image format authors.

> I chose to go with the `csv`-ish structure, with the core as a subfolder of the top level. I think this makes sense as there are fewer changes required.

Sounds good to me.

**Comment by jamwaffles** [comment]
Good to know I'm on the right path. The crate code will use imports from `embedded-graphics-core`, but should the tests and doc examples use _only_ embedded-graphics exports?

**Comment by rfuest** [comment]
> The crate code will use imports from `embedded-graphics-core`, but should the tests...

I'm not sure about the tests. But if we only use `embedded-graphics` it would create a much less noisy PR.

> ... and doc examples use _only_ embedded-graphics exports?

Most doc examples should only use `embedded-graphics` exports. But examples that are meant to show how display drivers, image formats or text renderes are implemented would use `embedded-graphics-core` exports.

**Comment by rfuest** [comment]
But there might be a problem when we add examples to the types in the core crate. When they are reexported into the main e-g docs, I would expect them to use `embedded-graphics` imports, but these wouldn't be available when the core crate docs are built.

**Comment by jamwaffles** [comment]
I've probably missed a load of things, but not really sure what they are. A cursory look at the structure and stuff would be good. I think the code itself is pretty ordinary as nothing's really changed - just lots of moving around.

**Comment by rfuest** [comment]
Looks good on the first cursory browse through the code!

Should we keep the structure identical between the crates? I think the only type that currently has a different location is `Rectangle`. Moving it into an, otherwise empty, primitives module wouldn't make much sense on it's own, but it would replicate the paths used in `embedded-graphics`. This might make migrating between `embedded-graphics` and `embedded-graphics-core` as easy as replacing all `use crate::embedded_graphics::`s with `use crate::embedded_graphics_core::`.

**Comment by jamwaffles** [comment]
I'd prefer to keep it as a single file. Having a `primitives/` folder suggests more will be added in the future, which is unlikely. True, it is more effort during migration, but not a lot I'd argue.

**Comment by jamwaffles** [comment]
I've done some moving around of the newly added methods from #500 in 5e31eb24ef2c03777c3c2f526e91b3edce2d46d8. This only leaves `center_offset` as a public method on `Size`. We could quite easily move this into a private function in `e-g` as I think it's utility outside of e-g is limited, but I'm interested to hear your thoughts @rfuest

**Comment by jamwaffles** [comment]
Ah, I forgot to mention: e-g-core needs it too! I could make it `pub` and optionally add a `#[doc(hidden])`, or we could just copy and paste the functions between crates (preferably not)

**Comment by rfuest** [comment]
We could move the code from `center_offset` into `Rectangle::with_center` and use `Rectangle::with_center(center, size).top_left` instead of `center_offset`.

**Comment by jamwaffles** [comment]
Good shout, thanks. 

There are a few bits of docs that are still TODO. Is there anything in particular you'd like to see? I'll get some stuff together and then I think this is ready for review. I'll convert to draft for now - I requested review too early.

**Comment by rfuest** [comment]
> Is there anything in particular you'd like to see?

No, nothing comes to my mind.

File: embedded-graphics-core/src/lib.rs
Comment (jamwaffles) [inline embedded-graphics-core/src/lib.rs:3]: @rfuest I could use a hand here. Bullet points are fine, but I'm drawing a blank at the moment.

Comment (jamwaffles) [reply embedded-graphics-core/src/lib.rs]: I had a cup of tea and the caffeine whispered sweet nothings into my brain. I've written some filler. Not sure if it's really in the right direction, but it's something we can work on.

File: embedded-graphics-core/src/rectangle/mod.rs
Comment (rfuest) [inline embedded-graphics-core/src/rectangle/mod.rs:450]: Should we move this type into the `geometry` module? It might be reused for other primitives in the future.

File: justfile
Comment (rfuest) [inline justfile:83]: `_build-readme` had an argument before the other crates were split from the e-g repo: https://github.com/embedded-graphics/embedded-graphics/blob/906ee1180cf31872b03aee037b9022c5fde30630/justfile#L111

File: tests/core-draw-target-cropped.rs
Comment (rfuest) [inline tests/core-draw-target-cropped.rs:1]: Why did you move these tests to here?

File: embedded-graphics-core/src/geometry/mod.rs
Comment (rfuest) [inline embedded-graphics-core/src/geometry/mod.rs:15]: This needs to be updated. `Dimensions` is also used by `DrawTarget`, where this description doesn't make much sense.

IMO this is one of the places were we need to consider the different target audiences for the docs. One the one hand we have normal users who need to known how to use `Dimensions` to get the dimensions of drawable and draw targets. On the other hand we have authors of third party crates who need to know when and how the need to implement `Dimensions` and `OriginDimensions`. I think this is an important distinction and should also be considered for similar traits. I would suggest to put the user facing docs first and than add a common header like `# Implementing Dimensions` or `# Implementation notes` followed by the third party crate developer docs. I've started to follow this pattern in the docs for `ImageDrawable` (maybe not the greatest example).

**Review by rfuest — CHANGES_REQUESTED** [review]
Here is the result of my first review pass.

File: .circleci/Dockerfile
Comment (jamwaffles) [reply .circleci/Dockerfile:8]: Yeah that would be good. I only added this comment because I saw the MSRV in the `deadlinks` readme. I did have a quick look at the `just` repo but couldn't find any versioning info there.

File: MIGRATING-0.6-0.7.md
Comment (jamwaffles) [reply MIGRATING-0.6-0.7.md:102]: Are the `For crates that handle images` and `For text rendering crates` sections enough or do you mean I should add more content to this `For display driver authors` section?

File: tests/core-draw-target-cropped.rs
Comment (jamwaffles) [reply tests/core-draw-target-cropped.rs:1]: IIRC it was to keep the e-g dependency out of core. It should be possible to add it as a dev dependency and move these tests back into their respective files, but I didn't do that in case it causes more dep conflicts in the future. What would you suggest?

File: embedded-graphics-core/src/geometry/size.rs
Comment (jamwaffles) [reply embedded-graphics-core/src/geometry/size.rs]: I've added it to both

File: embedded-graphics-core/src/lib.rs
Comment (jamwaffles) [reply embedded-graphics-core/src/lib.rs]: I'd like to create a core variant at some point, but we can do that later.

File: embedded-graphics-core/src/geometry/mod.rs
Comment (jamwaffles) [reply embedded-graphics-core/src/geometry/mod.rs]: I like the `implementation notes` idea, although I could use some suggestions for what goes above that heading, and below it.

File: embedded-graphics-core/src/lib.rs
Comment (jamwaffles) [reply embedded-graphics-core/src/lib.rs]: We need `Drawable` and `Pixel` to match e-g, but I've removed `Rectangle`.

File: embedded-graphics-core/src/lib.rs
Comment (jamwaffles) [reply embedded-graphics-core/src/lib.rs]: I've copied and pasted it

**Comment by jamwaffles** [comment]
My apologies if I've missed some comments - there were a lot, this PR is huge and it's Friday night :grimacing:

File: MIGRATING-0.6-0.7.md
Comment (rfuest) [reply MIGRATING-0.6-0.7.md:102]: That's enough. I just wanted to make sure we remember to mention the core crate when we add these sections.

File: tests/core-draw-target-cropped.rs
Comment (rfuest) [reply tests/core-draw-target-cropped.rs:1]: But doesn't the core crate already have a dev-dependency on e-g?

File: embedded-graphics-core/src/geometry/mod.rs
Comment (rfuest) [reply embedded-graphics-core/src/geometry/mod.rs]: Above:

* What objects the can this trait be used on.
* A description of which dimension is defined by this trait: For styled primitives it's the bounding box around all drawn pixels, but for draw target it's the drawable area.
* The difference between the dimensions of styled and unstyled primitives.
* Maybe some suggestions how the result can be used, like being able to use `.bounding_box().points()`.
* At some point we might also want to add examples to this, but that's even more out of scope of this PR than the rest of this suggestion :wink:.

Below:

* That some traits require this trait to be implemented, perhaps with a list of other traits.
* Your description of `OriginDimensions` isn't accurate: The `top_left` is also known and is guaranteed to be `Point::zero()`.
* Maybe a suggestion that the impl of `Dimensions` for drawables can be tested by using `MockDisplay::affeted_area`.

File: tests/core-draw-target-cropped.rs
Comment (jamwaffles) [reply tests/core-draw-target-cropped.rs:1]: Yes it does :facepalm:. Not sure where I was going with the tests moved here, but I can move them back. With one strange quirk: all their dependencies must come from e-g to compile. I can't use a mix of `crate::` and `embedded_graphics::`. Unsure why, but they work now so :shrug:

Comment (rfuest) [reply tests/core-draw-target-cropped.rs]: I think this is be caused by the cyclic dev-dependency. Cargo will have to break the cycle somewhere and this makes the types from `crate` and `embedded_graphics` incompatible. 

The cyclic dev-dependency will also cause a problem when we want to publish the crate, because we can't publish one create without the other already existing. But this can be fixed by removing the `version` from the dev-dependency in the core crate: See the note at https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#development-dependencies

Comment (jamwaffles) [reply tests/core-draw-target-cropped.rs]: Another TIL on Cargo's behaviour. Is it just me or is it really complex when you get into the weeds? Anyway, I've removed the version so now the dev dependency should hopefully not cause any problems.

**Comment by jamwaffles** [comment]
When this PR is merged, I want to release core onto crates.io. Should we release as `1.0.0` or `0.1.0`? If we want to go for a 1.0.0 release, we could release `0.1.0` now to make sure everything works ok, then release 1.0.0 when e-g 0.7 is cut. That gives us time to make breaking changes before committing to a major version. 

We could also just release core as < 1.0.0 for e-g 0.7. Thoughts?

File: core/src/primitives/rectangle/points.rs
Comment (rfuest) [inline core/src/primitives/rectangle/points.rs:2]: IMO we don't need to add doc comments to private modules, because they will never be seen in to docs.
```suggestion
```

File: src/geometry/mod.rs
Comment (rfuest) [inline src/geometry/mod.rs:24]: The doc comments should be moves to `PointExt`, otherwise they don't show up when using code completion.

File: src/lib.rs
Comment (rfuest) [inline src/lib.rs:306]: If we duplicate the trait I would also duplicate the test.

File: CHANGELOG.md
Comment (rfuest) [inline CHANGELOG.md:20]: There should also be an entry in the changelog(s) that `Primitive::points` no longer exists and was replaced by `PointsIter::points`.

**Review by rfuest — CHANGES_REQUESTED** [review]
Here are some more comments

**Comment by rfuest** [comment]
> We could also just release core as < 1.0.0 for e-g 0.7. Thoughts?

I'm not sure yet. I would suggest we decide this when it's time to release 0.7. Until then we can release core as 0.1.0, as you suggested.

File: core/src/primitives/rectangle/mod.rs
Comment (jamwaffles) [reply core/src/primitives/rectangle/mod.rs:433]: It was indeed. Should we also move `PointIter` out of the `primitives` module and into `core/src/iterator/points.rs`?

**Comment by jamwaffles** [comment]
> I'm not sure yet. I would suggest we decide this when it's time to release 0.7. Until then we can release core as 0.1.0, as you suggested.

SGTM.

**Review by rfuest — APPROVED** [review]
There are still some thing I'd like to tweak before the release, but I would suggest to merge this PR first and handle all further changes in smaller PRs that are easier to review.

Changes I would like to make or at least try out:
* Moving the `PointsIter` trait into the `iterator` module seems like a good idea, but I think the `iterator` module could use a cleanup anyway and perhaps some types could be moved out of the core crate.
* I would like to remove `RawDataIter` from `core`, because it's more likely to change than the other stuff in the `pixelcolor` module. Perhaps it can also be moved into `iterator`.
* We currently need to include the draw target adapters, like `Clipped` or `ColorConverted` in `core`, but IMO they don't really belong there. But I think there are several ways to fix this, which we can discuss later.

**Comment by jamwaffles** [comment]
Thanks for the reviews! Those points make sense. Please make an issue if you think it's worth tracking.

With those changes in mind, should I hold off on a `0.1.0` release of core?

**Comment by rfuest** [comment]
It might be a good idea to release `0.1.0` now, so we can start testing it with other crates. The suggested changes will be breaking, but as long as the core crate isn't widely adopted there is no harm in breaking the API a few times until the `0.7` release.

Do you also want to release a new e-g alpha that is compatible with core `0.1.0`?

**Comment by jamwaffles** [comment]
Yep, I'll get both crates released. Hopefully it goes smoothly!

**Comment by rfuest** [comment]
> Hopefully it goes smoothly!

:crossed_fingers:

**Comment by jamwaffles** [comment]
core 0.1.0 is released: https://crates.io/crates/embedded-graphics-core
e-g 0.7.0-alpha.2 is released: https://crates.io/crates/embedded-graphics/0.7.0-alpha.2

`cargo-release` didn't update the changelogs which is annoying. I'll fix those now.

**Comment by jamwaffles** [comment]
Everything is fixed. I force pushed to master as the release commits/tags generated by `cargo-release` were incorrect. Everything should be good now. I haven't gone through the checklist in #382 but I've created some GH releases, and I'll throw some announcements in the e-g Matrix chat.

**Pull request merged**
